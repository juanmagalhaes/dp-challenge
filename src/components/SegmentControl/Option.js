import styled from "styled-components";
import PropTypes from "prop-types";

import { colors, spacing } from "app/styles";

const Option = styled.span`
  background-color: ${({ dark }) => (dark ? colors.elm : colors.white)};
  border-radius: ${spacing.px.small};
  box-sizing: border-box;
  color: ${({ dark }) => (dark ? colors.white : colors.jellyBean)};
  cursor: pointer;
  font-size: 1.3rem;
  padding: ${spacing.px.large} ${spacing.px.larger};
  text-align: center;
  user-select: none;

  ${({ small }) =>
    small &&
    `
    padding: ${spacing.px.medium} ${spacing.px.large};
  `}

  ${({ width }) => width && `flex: 1;`}

  ${({ active, dark }) =>
    active &&
    `
    background-color: ${dark ? colors.white : colors.jellyBean};
    color: ${dark ? colors.jellyBean : colors.white};
  `}

  ${({ mini, active, dark }) =>
    mini &&
    `
    padding: ${spacing.px.small} ${spacing.px.medium};
    font-size: 1rem;

    ${!active &&
      !dark &&
      `
      background: ${colors.catskillWhite};
    `}
  `}

`;

Option.propTypes = {
  dark: PropTypes.bool,
  mini: PropTypes.bool,
  active: PropTypes.bool,
  small: PropTypes.bool,
  width: PropTypes.string
};

export default Option;
