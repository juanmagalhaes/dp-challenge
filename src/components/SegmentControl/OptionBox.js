import React from "react";
import styled from "styled-components";

import { colors, spacing } from "app/styles";

const Box = {};

Box.Inner = styled.div`
  background-color: ${({ dark }) => (dark ? colors.elm : colors.white)};
  border: solid 1px ${({ dark }) => (dark ? colors.elm : colors.jellyBean)};
  border-radius: ${spacing.px.medium};
  display: inline-flex;
  padding: ${spacing.px.small};
  width: 100%;

  ${({ small }) =>
    small &&
    `
    padding: ${spacing.px.medium};
  `} ${({ mini, dark }) =>
    mini &&
    `
    background: ${colors.catskillWhite};
    border-radius: ${spacing.px.small};
    padding: 0;

    ${dark &&
      `
      background: ${colors.elm};
    `}
  `};
`;

Box.Outer = styled.div`
  background-color: ${({ dark }) => (dark ? colors.jellyBean : colors.white)};
  display: inline-flex;
  padding: ${spacing.px.large};
  ${({ width }) =>
    width &&
    `
    width: ${width};
  `}

  ${({ small }) =>
    small &&
    `
    padding: ${spacing.px.medium};
  `}

  ${({ mini }) =>
    mini &&
    `
    padding: ${spacing.px.medium};
  `}
`;

export default ({ children, ...rest }) => (
  <Box.Outer {...rest}>
    <Box.Inner {...rest}>{children}</Box.Inner>
  </Box.Outer>
);
