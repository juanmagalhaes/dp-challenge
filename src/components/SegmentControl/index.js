export { default } from "./Container";
export { default as Option } from "./Option";
