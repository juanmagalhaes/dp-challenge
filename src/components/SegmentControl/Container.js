import { cloneElement } from "react";
import {
  branch,
  compose,
  mapProps,
  renderNothing,
  setDisplayName,
  setPropTypes,
  setStatic,
  withState,
  withHandlers
} from "recompose";
import PropTypes from "prop-types";

import SegmentControl from "./SegmentControl";
import Option from "./Option";

const propTypes = {
  children: PropTypes.node.isRequired,
  dark: PropTypes.bool,
  mini: PropTypes.bool,
  onChange: PropTypes.func,
  selected: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  small: PropTypes.bool,
  width: PropTypes.string
};

export default compose(
  setPropTypes(propTypes),
  setDisplayName("SegmentControl"),
  setStatic("Option", Option),
  branch(({ children }) => !children, renderNothing),
  withState("active", "setActive", ({ children }) => {
    if (children.length) {
      const selectedChildren = children.findIndex(
        child => child.props.selected
      );
      if (selectedChildren !== -1) {
        return selectedChildren;
      }
    }

    return 0;
  }),
  withHandlers({
    onChildClick: ({ setActive }) => childIndex => setActive(childIndex)
  }),
  mapProps(({ children, ...props }) => {
    const enhanceChild = (child, index = 0) => {
      return cloneElement(child, {
        dark: props.dark,
        key: index,
        onClick: evt => {
          props.onChildClick(index);
          if (props.onChange) {
            props.onChange({ ...evt, type: "change" }, index);
          }
          if (child.props.onClick) {
            child.props.onClick(evt);
          }
        },
        mini: props.mini,
        selected: props.active === index,
        small: props.small,
        width: props.width,
        ...child.props
      });
    };

    return {
      ...props,
      children: !children.length
        ? enhanceChild(children)
        : children.map(enhanceChild)
    };
  })
)(SegmentControl);
