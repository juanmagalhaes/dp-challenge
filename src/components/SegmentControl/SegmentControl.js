import React from "react";

import OptionBox from "./OptionBox";

const SegmentControl = ({ children, ...rest }) => (
  <OptionBox {...rest}>{children}</OptionBox>
);

export default SegmentControl;
