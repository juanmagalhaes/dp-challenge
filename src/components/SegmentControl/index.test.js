import React from "react";
import SegmentControl, { Option } from "./";

describe("Question 7 => SegmentControl", () => {
  test("render with 2 children", () => {
    const component = shallow(
      <SegmentControl>
        <Option>Option 1</Option>
        <Option>Option 2</Option>
      </SegmentControl>
    );

    expect(component).toMatchSnapshot();
  });

  test("render with several children", () => {
    const component = shallow(
      <SegmentControl>
        <Option>Option 1</Option>
        <Option>Option 2</Option>
        <Option>Option 3</Option>
        <Option>Option 4</Option>
        <Option>Option 5</Option>
      </SegmentControl>
    );

    expect(component).toMatchSnapshot();
  });

  test("render with dark theme", () => {
    const component = shallow(
      <SegmentControl dark>
        <Option>Option 1</Option>
        <Option>Option 2</Option>
      </SegmentControl>
    );

    expect(component).toMatchSnapshot();
  });

  test("render with small prop", () => {
    const component = shallow(
      <SegmentControl small>
        <Option>Option 1</Option>
        <Option>Option 2</Option>
      </SegmentControl>
    );

    expect(component).toMatchSnapshot();
  });

  test("render with mini prop", () => {
    const component = shallow(
      <SegmentControl mini>
        <Option>Option 1</Option>
        <Option>Option 2</Option>
      </SegmentControl>
    );

    expect(component).toMatchSnapshot();
  });

  test("render with Option selected", () => {
    const component = shallow(
      <SegmentControl>
        <Option>Option 1</Option>
        <Option selected>Option 2</Option>
      </SegmentControl>
    );

    expect(component).toMatchSnapshot();
  });

  test("change handler", () => {
    const handlerMock = jest.fn();
    const component = shallow(
      <SegmentControl onChange={handlerMock}>
        <Option>Option 1</Option>
        <Option selected>Option 2</Option>
      </SegmentControl>
    );

    component.simulate("change");

    expect(handlerMock).toHaveBeenCalled();
  });

  test("change handler", () => {
    const handlerMock = jest.fn();
    const component = shallow(
      <SegmentControl onChange={handlerMock}>
        <Option>Option 1</Option>
        <Option>Option 2</Option>
      </SegmentControl>
    );

    expect(handlerMock).not.toHaveBeenCalled();
    component.simulate("change");
    expect(handlerMock).toHaveBeenCalled();
  });

  test("click handler", () => {
    const handlerMock = jest.fn();
    const component = shallow(
      <SegmentControl>
        <Option selected>Option 1</Option>
        <Option onClick={handlerMock}>Option 2</Option>
      </SegmentControl>
    );

    const option2 = component.find(Option).last();
    expect(handlerMock).not.toHaveBeenCalled();
    option2.simulate("click");
    expect(handlerMock).toHaveBeenCalled();
  });
});
