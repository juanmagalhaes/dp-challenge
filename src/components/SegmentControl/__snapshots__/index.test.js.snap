// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Question 7 => SegmentControl render with 2 children 1`] = `
<withState(withHandlers(mapProps(SegmentControl)))>
  <Option>
    Option 1
  </Option>
  <Option>
    Option 2
  </Option>
</withState(withHandlers(mapProps(SegmentControl)))>
`;

exports[`Question 7 => SegmentControl render with Option selected 1`] = `
<withState(withHandlers(mapProps(SegmentControl)))>
  <Option>
    Option 1
  </Option>
  <Option
    selected={true}
  >
    Option 2
  </Option>
</withState(withHandlers(mapProps(SegmentControl)))>
`;

exports[`Question 7 => SegmentControl render with dark theme 1`] = `
<withState(withHandlers(mapProps(SegmentControl)))
  dark={true}
>
  <Option>
    Option 1
  </Option>
  <Option>
    Option 2
  </Option>
</withState(withHandlers(mapProps(SegmentControl)))>
`;

exports[`Question 7 => SegmentControl render with mini prop 1`] = `
<withState(withHandlers(mapProps(SegmentControl)))
  mini={true}
>
  <Option>
    Option 1
  </Option>
  <Option>
    Option 2
  </Option>
</withState(withHandlers(mapProps(SegmentControl)))>
`;

exports[`Question 7 => SegmentControl render with several children 1`] = `
<withState(withHandlers(mapProps(SegmentControl)))>
  <Option>
    Option 1
  </Option>
  <Option>
    Option 2
  </Option>
  <Option>
    Option 3
  </Option>
  <Option>
    Option 4
  </Option>
  <Option>
    Option 5
  </Option>
</withState(withHandlers(mapProps(SegmentControl)))>
`;

exports[`Question 7 => SegmentControl render with small prop 1`] = `
<withState(withHandlers(mapProps(SegmentControl)))
  small={true}
>
  <Option>
    Option 1
  </Option>
  <Option>
    Option 2
  </Option>
</withState(withHandlers(mapProps(SegmentControl)))>
`;
